common/src/main/java/com/microsoft/identity/common/internal/cache/ActiveBrokerCache.kt

```diff
@@ -0,0 +1,137 @@
+         **/
+        private const val BROKER_METADATA_CACHE_STORE_ON_BROKER_SIDE_STORAGE_NAME = "BROKER_METADATA_CACHE_STORE_ON_BROKER_SIDE"
+
+        /**
+         * File name of [ActiveBrokerCache] used by the SDK code.
+         **/
+        private const val BROKER_METADATA_CACHE_STORE_ON_SDK_SIDE_STORAGE_NAME = "BROKER_METADATA_CACHE_STORE_ON_SDK_SIDE"
+
+        /**
+         * The Lock for all [ActiveBrokerCache] instances used by the broker code.
+         **/
+        private val sBrokerSideLock = ReentrantReadWriteLock()
+
+        /**
+         * The Lock for all [ActiveBrokerCache] instances used by the SDK code.
+         **/
+        private val sSdkSideLock = ReentrantReadWriteLock()
+
+        /**
+         * If the caller is the broker, invoke this function.
+         *
+         * @param storageSupplier an [IStorageSupplier] component.
+         * @return a thread-safe [IActiveBrokerCache].
+         * */
+        fun getBrokerMetadataStoreOnBrokerSide(storageSupplier: IStorageSupplier)
+            : IActiveBrokerCache {
+            return ActiveBrokerCache(
+                storage = storageSupplier.getNameValueStore(
+                    BROKER_METADATA_CACHE_STORE_ON_BROKER_SIDE_STORAGE_NAME, String::class.java),
+                lock = sBrokerSideLock
+            )
+        }
+
+        /**
+         * If the caller is an SDK, invoke this function.
+         *
+         * @param storageSupplier an [IStorageSupplier] component.
+         * @return a thread-safe [IActiveBrokerCache].
+         */
+        fun getBrokerMetadataStoreOnSdkSide(storageSupplier: IStorageSupplier)
+                : IActiveBrokerCache {
+            return ActiveBrokerCache(
+                storage = storageSupplier.getNameValueStore(
+                    BROKER_METADATA_CACHE_STORE_ON_SDK_SIDE_STORAGE_NAME, String::class.java),
+                lock = sSdkSideLock
+            )
+        }
+    }
+
+    /**
+     * In-memory cached value.
+     */
+    internal var inMemoryCachedValue: BrokerData? = null
+
+    override fun getCachedActiveBroker(): BrokerData? {
+        lock.read {
+            if (inMemoryCachedValue != null) {
+                return inMemoryCachedValue
+            }
+
+            val packageName = storage.get(ACTIVE_BROKER_CACHE_PACKAGE_NAME_KEY)
+            val signatureHash = storage.get(ACTIVE_BROKER_CACHE_SIGHASH_KEY)
+
+            if (packageName.isNullOrEmpty() || signatureHash.isNullOrEmpty())
+                return null
+
+            inMemoryCachedValue = BrokerData(packageName, signatureHash)
+            return inMemoryCachedValue
+        }
+    }
+
+    override fun setCachedActiveBroker(brokerData: BrokerData) {
+        lock.write {
+            storage.put(ACTIVE_BROKER_CACHE_PACKAGE_NAME_KEY, brokerData.packageName)
+            storage.put(ACTIVE_BROKER_CACHE_SIGHASH_KEY, brokerData.signatureHash)
+            inMemoryCachedValue = BrokerData(brokerData.packageName, brokerData.signatureHash)
```

Comment:
BrokerData.copy() will also work? this is just the way you prefer to do it, right?

Reply:
lol no. You overestimated my skill 😋